File: cpp/tests/internal/fpga_info.cpp
// File: fpga_info.cpp
// Project: internal
// Created Date: 26/09/2023
// Author: Shun Suzuki
// -----
// Last Modified: 06/12/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2023 Shun Suzuki. All rights reserved.
//

#include <gtest/gtest.h>

#include <autd3/datagram/reads_fpga_info.hpp>
#include <autd3/internal/datagram.hpp>
#include <autd3/internal/exception.hpp>
#include <autd3/internal/fpga_info.hpp>

#include "utils.hpp"

TEST(Internal, FPGAInfo) {
  auto autd = create_controller();

  ASSERT_TRUE(autd.send_async(autd3::datagram::ConfigureReadsFPGAInfo([](const auto&) { return true; })).get());
  {
    for (const auto infos = autd.fpga_info_async().get(); auto info : infos) ASSERT_FALSE(info.is_thermal_assert());
  }

  {
    autd.link().assert_thermal_sensor(0);
    autd.link().update(0);
    autd.link().update(1);

    const auto infos = autd.fpga_info_async().get();
    ASSERT_TRUE(infos[0].is_thermal_assert());
    ASSERT_FALSE(infos[1].is_thermal_assert());
  }

  {
    autd.link().deassert_thermal_sensor(0);
    autd.link().assert_thermal_sensor(1);
    autd.link().update(0);
    autd.link().update(1);

    const auto infos = autd.fpga_info_async().get();
    ASSERT_FALSE(infos[0].is_thermal_assert());
    ASSERT_TRUE(infos[1].is_thermal_assert());
  }

  {
    autd.link().break_down();
    ASSERT_THROW((void)autd.fpga_info_async().get(), autd3::internal::AUTDException);
  }
}

File: src/autd3-driver/src/fpga/defined/fpga_info.rs
/*
 * File: fpga_info.rs
 * Project: defined
 * Created Date: 05/10/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 05/10/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 *
 */

use crate::cpu::RxMessage;

/// FPGA information
#[repr(C)]
pub struct FPGAInfo {
    info: u8,
}

impl FPGAInfo {
    pub const fn new(info: u8) -> Self {
        Self { info }
    }

    /// Check if thermal sensor is asserted
    pub const fn is_thermal_assert(&self) -> bool {
        (self.info & 0x01) != 0
    }

    pub const fn info(&self) -> u8 {
        self.info
    }
}

impl From<&RxMessage> for FPGAInfo {
    fn from(msg: &RxMessage) -> Self {
        Self { info: msg.data }
    }
}

#[cfg(test)]
mod tests {
    use std::mem::size_of;

    use super::*;

    #[test]
    fn fpga_info() {
        assert_eq!(size_of::<FPGAInfo>(), 1);

        let info = FPGAInfo::new(0x00);
        assert!(!info.is_thermal_assert());
        assert_eq!(info.info(), 0x00);

        let info = FPGAInfo::new(0x01);
        assert!(info.is_thermal_assert());
        assert_eq!(info.info(), 0x01);

        let info = FPGAInfo::new(0x02);
        assert!(!info.is_thermal_assert());
        assert_eq!(info.info(), 0x02);

        let rx = RxMessage { ack: 0, data: 0x01 };
        let info = FPGAInfo::from(&rx);
        assert!(info.is_thermal_assert());
        assert_eq!(info.info(), 0x01);
    }
}
